File: codility/lesson1/BinaryGap.java
/**
 * @author Igor Baiborodine
 *
 * @see ../BINARY_GAP_README.md
 */
public class BinaryGap {

  public int solution(int N) {

    int maxBinaryGap = 0;
    int currentBinaryGap = 0;
    int previousRemainder = -1;

    while (N != 0) {
      int currentRemainder = N % 2;

      if (currentRemainder == 0) {
        if (previousRemainder == 1) { // started new binary gap
          currentBinaryGap = 1;
        } else if (previousRemainder == 0
            && currentBinaryGap > 0) { // continue current binary gap
          currentBinaryGap++;
        }
      } else { // currentRemainder == 1
        if (previousRemainder == 0) { // completed current binary gap
          if (currentBinaryGap > maxBinaryGap) {
            maxBinaryGap = currentBinaryGap;
            currentBinaryGap = 0;
          }
        }
      }
      previousRemainder = currentRemainder;
      N /= 2;
    }

    return maxBinaryGap;
  }

}

File: codility/lesson1/BinaryGap.py
"""Solution for the Binary Gap exercise.

Exercise: BINARY_GAP_README.md
Author:   Igor Baiborodine

"""

def solution(N):
    max_bin_gap = 0
    cur_bin_gap = 0
    prev_rmnd = -1
    
    while (N != 0):
        cur_rmnd = N % 2
        
        if cur_rmnd == 0:
            if prev_rmnd == 1:
                cur_bin_gap = 1
            elif prev_rmnd == 0 and cur_bin_gap > 0:
                cur_bin_gap += 1
        else:
            if prev_rmnd == 0 and cur_bin_gap > max_bin_gap:
                max_bin_gap = cur_bin_gap
                cur_bin_gap = 0
          
        prev_rmnd = cur_rmnd
        N //= 2
    return max_bin_gap  
    
assert solution(5) == 1
assert solution(5728) == 2
assert solution(9) == 2
assert solution(529) == 4
assert solution(20) == 1
assert solution(15) == 0
assert solution(1041) == 5
